sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala

```diff
@@ -137,6 +137,22 @@ object RowEncoder {
         CreateStruct(convertedFields))
   }
 
+  private def externalDataTypeForInput(dt: DataType): DataType = dt match {
```

Comment:
I think we only need to handle `DecimalType` here, and call `externalDataTypeFor` for other cases.